uber/automated_emails.py

```diff
@@ -1,75 +1,290 @@
+        # on any external config like EVENT_NAME.  examples:
+        # good:  'registration_confirmation_email'
+        # bad:   '{EVENT_NAME}_confirmation_email'
+        assert '{EVENT_NAME}' not in self.ident, 'AutomatedEmail.ident "{}" not allowed to contain EVENT_NAME'.format(self.ident)
+        assert self.ident not in self.instances, 'error: ident "{}" is registered twice.'.format(self.ident)
+
+        self.instances[self.ident] = self
+
+        self.model, self.template, self.needs_approval = model, template, needs_approval
         self.cc = cc or []
         self.bcc = bcc or []
         self.extra_data = extra_data or {}
         self.sender = sender or c.REGDESK_EMAIL
-        self.instances[self.subject] = self
-        self.ident = (ident or self.subject).format(EVENT_NAME=c.EVENT_NAME)
+        self.when = listify(when)
+
         if post_con:
-            self.filter = lambda x: c.POST_CON and filter(x)
+            self.filter = lambda model_inst: c.POST_CON and filter(model_inst)
         else:
-            self.filter = lambda x: not c.POST_CON and filter(x)
+            self.filter = lambda model_inst: not c.POST_CON and filter(model_inst)
+
+    def filters_run(self, model_inst):
+        if self.filter and not self.filter(model_inst):
+            return False
+
+        for date_filter in self.when:
+            if not date_filter():
+                return False
+
+        return True
 
     def __repr__(self):
         return '<{}: {!r}>'.format(self.__class__.__name__, self.subject)
 
-    def prev(self, x, all_sent=None):
-        if all_sent:
-            return (x.__class__.__name__, x.id, self.ident) in all_sent
-        else:
-            with Session() as session:
-                return session.query(Email).filter_by(model=x.__class__.__name__, fk_id=x.id, ident=self.ident).first()
+    def _already_sent(self, model_inst):
+        """
+        Returns true if we have a record of previously sending this email to this model
+
+        NOTE: c.PREVIOUSLY_SENT_EMAILS is a cached property and will only update at the start of each daemon run.
+        """
+        return (model_inst.__class__.__name__, model_inst.id, self.ident) in c.PREVIOUSLY_SENT_EMAILS
 
-    def should_send(self, x, all_sent=None):
+    def send_if_possible(self, model_inst, raise_errors=False):
+        """
+        If it's OK to send an email of our category to this model instance (i.e. a particular Attendee) then send it.
+        """
         try:
-            return x.email and not self.prev(x, all_sent) and self.filter(x)
+            if self._should_send(model_inst):
+                self.really_send(model_inst)
         except:
-            log.error('unexpected error', exc_info=True)
+            log.error('error sending {!r} email to {}', self.subject, model_inst.email, exc_info=True)
+            if raise_errors:
+                raise
+
+    def _should_send(self, model_inst):
```

Comment:
I don't think _should_send can be a property because it takes an argument (i.e. an Attendee) that is different on each run.